File: v0.2/golang/pic/verifier.go
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

package pic

import (
	"fmt"
	"strings"
	"time"
)

// Verifier validates PCAs before any authority is exercised (Prover/Verifier
// spec §3). It resolves keys through a Registry, consults an optional
// RevocationStore, and keeps per-Verifier state for single-use challenges.
type Verifier struct {
	Registry    *Registry
	Revocations *RevocationStore // optional; nil means no revocation state consulted

	used map[string]bool // consumed single-use challenges (§6.1)
}

// NewVerifier returns a Verifier over the given key registry and (optional)
// revocation store.
func NewVerifier(reg *Registry, rev *RevocationStore) *Verifier {
	return &Verifier{Registry: reg, Revocations: rev, used: map[string]bool{}}
}

// VerifyOrigin validates a PCA0 (§3.2) and the revocation-coordinate derivation
// (Revocation spec §2.1, §2.4): signature under the issuer, validity window,
// profile present, and lineageId/branchId recomputed from the origin core.
func (v *Verifier) VerifyOrigin(p *PCA, now time.Time) error {
	if !p.IsOrigin() {
		return fmt.Errorf("origin validation: PCA carries a Proof of Relationship")
	}
	if p.Proof == nil {
		return fmt.Errorf("origin validation: missing signature")
	}
	msg, err := p.signingBytes()
	if err != nil {
		return err
	}
	if err := v.Registry.verify(p.Issuer, msg, p.Proof.Signature); err != nil {
		return fmt.Errorf("origin validation: %w", err)
	}
	if err := withinValidity(p.IssuedAt, p.ExpiresAt, now); err != nil {
		return fmt.Errorf("origin validation: %w", err)
	}
	if p.Profile != RevocableProfile {
		return fmt.Errorf("origin validation: unknown profile %q", p.Profile)
	}
	if p.LineageCounter != 0 {
		return fmt.Errorf("origin validation: lineageCounter must be 0, got %d", p.LineageCounter)
	}
	wantLineage, err := deriveLineageID(p)
	if err != nil {
		return err
	}
	if p.LineageID != wantLineage {
		return fmt.Errorf("origin validation: lineageId does not match origin commitment")
	}
	if p.BranchID != rootBranchID(p.LineageID) {
		return fmt.Errorf("origin validation: branchId is not the derived root branch id")
	}
	if v.Revocations != nil {
		if err := v.Revocations.Check(p); err != nil {
			return fmt.Errorf("origin validation: %w", err)
		}
	}
	return nil
}

// VerifyHop validates a non-origin PCA against its already-validated predecessor,
// performing the ordered checks of §3.3 plus the revocation-coordinate
// continuity of Revocation spec §2.3. If consume is true, single-use challenges
// are marked consumed (live acceptance); re-validation of history passes false.
func (v *Verifier) VerifyHop(cur, pred *PCA, now time.Time, consume bool) error {
	if cur.IsOrigin() {
		return fmt.Errorf("hop validation: PCA carries no Proof of Relationship")
	}
	por := cur.ProofOfRelationship

	// 1. integrity — single signature over the whole PCA.
	if cur.Proof == nil {
		return fmt.Errorf("hop: missing signature")
	}
	msg, err := cur.signingBytes()
	if err != nil {
		return err
	}
	if err := v.Registry.verify(cur.Proof.VerificationMethod, msg, cur.Proof.Signature); err != nil {
		return fmt.Errorf("hop integrity: %w", err)
	}

	// 2. predecessor binding — previousPcaHash equals the presented predecessor.
	predDigest, err := pred.Digest()
	if err != nil {
		return err
	}
	if por.PreviousPcaHash != predDigest {
		return fmt.Errorf("hop binding: previousPcaHash does not match the presented predecessor")
	}

	// Revocation-coordinate continuity (Revocation spec §2.3).
	if err := coordinateContinuity(cur, pred); err != nil {
		return fmt.Errorf("hop coordinates: %w", err)
	}

	// 3. continuation — response carries the predecessor challenge, unexpired,
	//    and (single-use) not already consumed.
	if por.ContinuationResponse.PredecessorChallenge != pred.Continuation.Challenge {
		return fmt.Errorf("hop continuation: response does not answer the predecessor challenge")
	}
	if !now.Before(pred.Continuation.ExpiresAt) {
		return fmt.Errorf("hop continuation: predecessor challenge expired")
	}
	if consume && pred.Continuation.Mode == "single-use" {
		if v.used[pred.Continuation.Challenge] {
			return fmt.Errorf("hop continuation: single-use challenge already consumed (replay)")
		}
	}

	// 4. attestation — embedded issuer signature valid, within validity, subject
	//    matches the executor, which matches the key that signed the PCA.
	if err := v.verifyAttestation(por, cur.Proof.VerificationMethod, now); err != nil {
		return fmt.Errorf("hop attestation: %w", err)
	}

	// 5. conformance — attested attributes satisfy the predecessor contract.
	if err := Conforms(por.ExecutorAttestation.Attributes, pred.Invariants.ExecutionContract); err != nil {
		return fmt.Errorf("hop conformance: %w", err)
	}

	// 6. non-expansion — invariants are equal to or more restrictive.
	if err := Attenuates(cur.Invariants, pred.Invariants); err != nil {
		return fmt.Errorf("hop non-expansion: %w", err)
	}

	// 7. temporal — hop window contained in the predecessor's (§6.3).
	if err := temporalCheck(cur, pred, now); err != nil {
		return fmt.Errorf("hop temporal: %w", err)
	}

	// (8. request match — executed-vs-signed — is enforced at execution time by
	//     the reference monitor against por.Request; see Authorize / scenario.)

	// revocation state — is this position cut off?
	if v.Revocations != nil {
		if err := v.Revocations.Check(cur); err != nil {
			return fmt.Errorf("hop revocation: %w", err)
		}
	}

	if consume && pred.Continuation.Mode == "single-use" {
		v.used[pred.Continuation.Challenge] = true
	}
	return nil
}

// coordinateContinuity enforces the hop-by-hop continuity of the revocation
// coordinates (Revocation spec §2.3).
func coordinateContinuity(cur, pred *PCA) error {
	if cur.Profile != pred.Profile {
		return fmt.Errorf("profile changed")
	}
	if cur.LineageID != pred.LineageID {
		return fmt.Errorf("lineageId changed")
	}
	if (cur.GrantID == "") != (pred.GrantID == "") || cur.GrantID != pred.GrantID {
		return fmt.Errorf("grantId presence/value changed")
	}
	if cur.OriginIssuer != pred.OriginIssuer {
		return fmt.Errorf("originIssuer changed")
	}
	if cur.BranchID != pred.BranchID {
		return fmt.Errorf("branchId changed without an authorized branch-creation transition")
	}
	// exact counter rule: current == predecessor + 1, no overflow/wrap.
	if cur.LineageCounter != pred.LineageCounter+1 {
		return fmt.Errorf("lineageCounter must be predecessor+1 (%d), got %d", pred.LineageCounter+1, cur.LineageCounter)
	}
	return nil
}

func (v *Verifier) verifyAttestation(por *PoR, pcaVM string, now time.Time) error {
	att := por.ExecutorAttestation
	if att.Proof == nil {
		return fmt.Errorf("attestation missing issuer signature")
	}
	msg, err := attestationSigningBytes(att)
	if err != nil {
		return err
	}
	if err := v.Registry.verify(att.Issuer, msg, att.Proof.Signature); err != nil {
		return fmt.Errorf("issuer signature: %w", err)
	}
	if err := withinValidity(att.IssuedAt, att.ExpiresAt, now); err != nil {
		return fmt.Errorf("attestation validity: %w", err)
	}
	if att.Subject != por.Executor {
		return fmt.Errorf("attestation subject %q does not match executor %q", att.Subject, por.Executor)
	}
	// the key that signed the PCA must belong to the executor.
	if !strings.HasPrefix(pcaVM, por.Executor) {
		return fmt.Errorf("PCA signing key %q does not belong to executor %q", pcaVM, por.Executor)
	}
	return nil
}

// VerifyFullChain validates a whole chain from PCA0 to the tip (Full Hash Chain
// profile, §5.1): cost O(n). It returns the invariants authorized at the tip.
// History re-validation does not consume single-use challenges.
func (v *Verifier) VerifyFullChain(chain []*PCA, now time.Time) (Invariants, error) {
	if len(chain) == 0 {
		return Invariants{}, fmt.Errorf("empty chain")
	}
	if err := v.VerifyOrigin(chain[0], now); err != nil {
		return Invariants{}, err
	}
	for i := 1; i < len(chain); i++ {
		if err := v.VerifyHop(chain[i], chain[i-1], now, false); err != nil {
			return Invariants{}, fmt.Errorf("hop %d: %w", i, err)
		}
	}
	return chain[len(chain)-1].Invariants, nil
}

// VerifyEnvelope validates one incremental transition carried in an envelope
// (§6.8): envelope signature and digests, then the single hop cur-against-pred,
// consuming the predecessor's single-use challenge. It trusts the predecessor
// inductively (validated by the hop that produced it) rather than walking back.
func (v *Verifier) VerifyEnvelope(env *Envelope, now time.Time) (Invariants, error) {
	body := env.Envelope
	if body.Current == nil || body.Predecessor == nil {
		return Invariants{}, fmt.Errorf("envelope: missing predecessor or current")
	}
	if env.Proof == nil {
		return Invariants{}, fmt.Errorf("envelope: missing signature")
	}
	msg, err := env.signingBytes()
	if err != nil {
		return Invariants{}, err
	}
	if err := v.Registry.verify(env.Proof.VerificationMethod, msg, env.Proof.Signature); err != nil {
		return Invariants{}, fmt.Errorf("envelope signature: %w", err)
	}
	// digests are convenience, not trusted input: recompute and cross-check.
	predDigest, err := body.Predecessor.Digest()
	if err != nil {
		return Invariants{}, err
	}
	curDigest, err := body.Current.Digest()
	if err != nil {
		return Invariants{}, err
	}
	if body.PredecessorDigest != predDigest || body.CurrentDigest != curDigest {
		return Invariants{}, fmt.Errorf("envelope: supplied digest does not match recomputed digest")
	}
	if body.Current.ProofOfRelationship.PreviousPcaHash != predDigest {
		return Invariants{}, fmt.Errorf("envelope: current.previousPcaHash does not equal predecessorDigest")
	}
	if v.Revocations != nil {
		if err := v.Revocations.Check(body.Predecessor); err != nil {
			return Invariants{}, fmt.Errorf("envelope predecessor: %w", err)
		}
	}
	if err := v.VerifyHop(body.Current, body.Predecessor, now, true); err != nil {
		return Invariants{}, err
	}
	return body.Current.Invariants, nil
}

func withinValidity(issuedAt, expiresAt, now time.Time) error {
	if now.Before(issuedAt) {
		return fmt.Errorf("not yet valid")
	}
	if !now.Before(expiresAt) {
		return fmt.Errorf("expired")
	}
	return nil
}

func temporalCheck(cur, pred *PCA, now time.Time) error {
	if cur.IssuedAt.Before(pred.IssuedAt) {
		return fmt.Errorf("issuedAt precedes predecessor")
	}
	if cur.ExpiresAt.After(pred.ExpiresAt) {
		return fmt.Errorf("expiresAt exceeds predecessor")
	}
	if err := withinValidity(cur.IssuedAt, cur.ExpiresAt, now); err != nil {
		return err
	}
	if cur.Continuation.ExpiresAt.After(pred.ExpiresAt) {
		return fmt.Errorf("emitted challenge outlives the lineage")
	}
	return nil
}

File: v0.2/rust/src/verifier.rs
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

//! The Verifier: origin validation, ordered per-hop checks (§3.3), full-chain and
//! envelope validation, plus the revocation-coordinate continuity of the
//! Revocation spec §2.3.

use crate::authority::{attenuates, conforms};
use crate::crypto::Registry;
use crate::revocation::{derive_lineage_id, root_branch_id};
use crate::types::{Envelope, Invariants, Pca, Por};
use crate::{parse_rfc3339, PicResult, RevocationStore, REVOCABLE_PROFILE};
use chrono::{DateTime, Utc};
use std::collections::HashSet;

/// Validates PCAs before any authority is exercised (Prover/Verifier spec §3). It
/// resolves keys through a Registry, consults an optional RevocationStore, and
/// keeps per-Verifier state for single-use challenges.
pub struct Verifier<'a> {
    pub registry: &'a Registry,
    pub revocations: Option<&'a RevocationStore>,
    used: HashSet<String>,
}

impl<'a> Verifier<'a> {
    /// Returns a Verifier over the given key registry and (optional) revocation
    /// store.
    pub fn new(registry: &'a Registry, revocations: Option<&'a RevocationStore>) -> Verifier<'a> {
        Verifier {
            registry,
            revocations,
            used: HashSet::new(),
        }
    }

    /// Validates a PCA0 (§3.2) and the revocation-coordinate derivation
    /// (Revocation spec §2.1, §2.4).
    pub fn verify_origin(&self, p: &Pca, now: DateTime<Utc>) -> PicResult<()> {
        if !p.is_origin() {
            return Err("origin validation: PCA carries a Proof of Relationship".to_string());
        }
        let proof = p
            .proof
            .as_ref()
            .ok_or("origin validation: missing signature")?;
        let msg = p.signing_bytes();
        self.registry
            .verify(&p.issuer, &msg, &proof.signature)
            .map_err(|e| format!("origin validation: {e}"))?;
        within_validity(&p.issued_at, &p.expires_at, now)
            .map_err(|e| format!("origin validation: {e}"))?;
        if p.profile != REVOCABLE_PROFILE {
            return Err(format!(
                "origin validation: unknown profile {:?}",
                p.profile
            ));
        }
        if p.lineage_counter != 0 {
            return Err(format!(
                "origin validation: lineageCounter must be 0, got {}",
                p.lineage_counter
            ));
        }
        let want_lineage = derive_lineage_id(p);
        if p.lineage_id != want_lineage {
            return Err(
                "origin validation: lineageId does not match origin commitment".to_string(),
            );
        }
        if p.branch_id != root_branch_id(&p.lineage_id) {
            return Err(
                "origin validation: branchId is not the derived root branch id".to_string(),
            );
        }
        if let Some(rev) = self.revocations {
            rev.check(p)
                .map_err(|e| format!("origin validation: {e}"))?;
        }
        Ok(())
    }

    /// Validates a non-origin PCA against its already-validated predecessor,
    /// performing the ordered checks of §3.3 plus revocation-coordinate
    /// continuity. If `consume` is true, single-use challenges are marked
    /// consumed (live acceptance); re-validation of history passes false.
    pub fn verify_hop(
        &mut self,
        cur: &Pca,
        pred: &Pca,
        now: DateTime<Utc>,
        consume: bool,
    ) -> PicResult<()> {
        if cur.is_origin() {
            return Err("hop validation: PCA carries no Proof of Relationship".to_string());
        }
        let por = cur.proof_of_relationship.as_ref().unwrap();

        // 1. integrity — single signature over the whole PCA.
        let proof = cur.proof.as_ref().ok_or("hop: missing signature")?;
        let msg = cur.signing_bytes();
        self.registry
            .verify(&proof.verification_method, &msg, &proof.signature)
            .map_err(|e| format!("hop integrity: {e}"))?;

        // 2. predecessor binding — previousPcaHash equals the presented predecessor.
        let pred_digest = pred.digest();
        if por.previous_pca_hash != pred_digest {
            return Err(
                "hop binding: previousPcaHash does not match the presented predecessor".to_string(),
            );
        }

        // Revocation-coordinate continuity (Revocation spec §2.3).
        coordinate_continuity(cur, pred).map_err(|e| format!("hop coordinates: {e}"))?;

        // 3. continuation — response carries the predecessor challenge, unexpired,
        //    and (single-use) not already consumed.
        if por.continuation_response.predecessor_challenge != pred.continuation.challenge {
            return Err(
                "hop continuation: response does not answer the predecessor challenge".to_string(),
            );
        }
        if now >= parse_rfc3339(&pred.continuation.expires_at) {
            return Err("hop continuation: predecessor challenge expired".to_string());
        }
        if consume
            && pred.continuation.mode == "single-use"
            && self.used.contains(&pred.continuation.challenge)
        {
            return Err(
                "hop continuation: single-use challenge already consumed (replay)".to_string(),
            );
        }

        // 4. attestation — embedded issuer signature valid, within validity,
        //    subject matches the executor, which matches the PCA signing key.
        self.verify_attestation(por, &proof.verification_method, now)
            .map_err(|e| format!("hop attestation: {e}"))?;

        // 5. conformance — attested attributes satisfy the predecessor contract.
        conforms(
            &por.executor_attestation.attributes,
            &pred.invariants.execution_contract,
        )
        .map_err(|e| format!("hop conformance: {e}"))?;

        // 6. non-expansion — invariants are equal to or more restrictive.
        attenuates(&cur.invariants, &pred.invariants)
            .map_err(|e| format!("hop non-expansion: {e}"))?;

        // 7. temporal — hop window contained in the predecessor's (§6.3).
        temporal_check(cur, pred, now).map_err(|e| format!("hop temporal: {e}"))?;

        // revocation state — is this position cut off?
        if let Some(rev) = self.revocations {
            rev.check(cur).map_err(|e| format!("hop revocation: {e}"))?;
        }

        if consume && pred.continuation.mode == "single-use" {
            self.used.insert(pred.continuation.challenge.clone());
        }
        Ok(())
    }

    fn verify_attestation(&self, por: &Por, pca_vm: &str, now: DateTime<Utc>) -> PicResult<()> {
        let att = &por.executor_attestation;
        let proof = att
            .proof
            .as_ref()
            .ok_or("attestation missing issuer signature")?;
        let msg = att.signing_bytes();
        self.registry
            .verify(&att.issuer, &msg, &proof.signature)
            .map_err(|e| format!("issuer signature: {e}"))?;
        within_validity(&att.issued_at, &att.expires_at, now)
            .map_err(|e| format!("attestation validity: {e}"))?;
        if att.subject != por.executor {
            return Err(format!(
                "attestation subject {:?} does not match executor {:?}",
                att.subject, por.executor
            ));
        }
        // the key that signed the PCA must belong to the executor.
        if !pca_vm.starts_with(&por.executor) {
            return Err(format!(
                "PCA signing key {pca_vm:?} does not belong to executor {:?}",
                por.executor
            ));
        }
        Ok(())
    }

    /// Validates a whole chain from PCA0 to the tip (Full Hash Chain profile,
    /// §5.1): cost O(n). Returns the invariants authorized at the tip. History
    /// re-validation does not consume single-use challenges.
    pub fn verify_full_chain(
        &mut self,
        chain: &[Pca],
        now: DateTime<Utc>,
    ) -> PicResult<Invariants> {
        if chain.is_empty() {
            return Err("empty chain".to_string());
        }
        self.verify_origin(&chain[0], now)?;
        for i in 1..chain.len() {
            self.verify_hop(&chain[i], &chain[i - 1], now, false)
                .map_err(|e| format!("hop {i}: {e}"))?;
        }
        Ok(chain[chain.len() - 1].invariants.clone())
    }

    /// Validates one incremental transition carried in an envelope (§6.8):
    /// envelope signature and digests, then the single hop cur-against-pred,
    /// consuming the predecessor's single-use challenge.
    pub fn verify_envelope(&mut self, env: &Envelope, now: DateTime<Utc>) -> PicResult<Invariants> {
        let body = &env.envelope;
        let (pred, cur) = match (&body.predecessor, &body.current) {
            (Some(p), Some(c)) => (p, c),
            _ => return Err("envelope: missing predecessor or current".to_string()),
        };
        let proof = env.proof.as_ref().ok_or("envelope: missing signature")?;
        let msg = env.signing_bytes();
        self.registry
            .verify(&proof.verification_method, &msg, &proof.signature)
            .map_err(|e| format!("envelope signature: {e}"))?;
        // digests are convenience, not trusted input: recompute and cross-check.
        let pred_digest = pred.digest();
        let cur_digest = cur.digest();
        if body.predecessor_digest != pred_digest || body.current_digest != cur_digest {
            return Err("envelope: supplied digest does not match recomputed digest".to_string());
        }
        let por = cur
            .proof_of_relationship
            .as_ref()
            .ok_or("envelope: current carries no Proof of Relationship")?;
        if por.previous_pca_hash != pred_digest {
            return Err(
                "envelope: current.previousPcaHash does not equal predecessorDigest".to_string(),
            );
        }
        if let Some(rev) = self.revocations {
            rev.check(pred)
                .map_err(|e| format!("envelope predecessor: {e}"))?;
        }
        self.verify_hop(cur, pred, now, true)?;
        Ok(cur.invariants.clone())
    }
}

/// Enforces the hop-by-hop continuity of the revocation coordinates (Revocation
/// spec §2.3).
fn coordinate_continuity(cur: &Pca, pred: &Pca) -> PicResult<()> {
    if cur.profile != pred.profile {
        return Err("profile changed".to_string());
    }
    if cur.lineage_id != pred.lineage_id {
        return Err("lineageId changed".to_string());
    }
    if (cur.grant_id.is_empty()) != (pred.grant_id.is_empty()) || cur.grant_id != pred.grant_id {
        return Err("grantId presence/value changed".to_string());
    }
    if cur.origin_issuer != pred.origin_issuer {
        return Err("originIssuer changed".to_string());
    }
    if cur.branch_id != pred.branch_id {
        return Err(
            "branchId changed without an authorized branch-creation transition".to_string(),
        );
    }
    if cur.lineage_counter != pred.lineage_counter + 1 {
        return Err(format!(
            "lineageCounter must be predecessor+1 ({}), got {}",
            pred.lineage_counter + 1,
            cur.lineage_counter
        ));
    }
    Ok(())
}

fn within_validity(issued_at: &str, expires_at: &str, now: DateTime<Utc>) -> PicResult<()> {
    let issued = parse_rfc3339(issued_at);
    let expires = parse_rfc3339(expires_at);
    if now < issued {
        return Err("not yet valid".to_string());
    }
    if now >= expires {
        return Err("expired".to_string());
    }
    Ok(())
}

fn temporal_check(cur: &Pca, pred: &Pca, now: DateTime<Utc>) -> PicResult<()> {
    if parse_rfc3339(&cur.issued_at) < parse_rfc3339(&pred.issued_at) {
        return Err("issuedAt precedes predecessor".to_string());
    }
    if parse_rfc3339(&cur.expires_at) > parse_rfc3339(&pred.expires_at) {
        return Err("expiresAt exceeds predecessor".to_string());
    }
    within_validity(&cur.issued_at, &cur.expires_at, now)?;
    if parse_rfc3339(&cur.continuation.expires_at) > parse_rfc3339(&pred.expires_at) {
        return Err("emitted challenge outlives the lineage".to_string());
    }
    Ok(())
}
